Add functionality to update owner public keys & threshold config in multisig

partylikeits1983 opened the pull request:
This PR adds the ability to update owner public keys and the threshold config in the multisig authentication component. 

Note: wait for #1883 & #1887 before merging.

Resolves #1707

partylikeits1983 commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +70,61 @@ proc.assert_new_tx
+        dup neq.0
+        # => [is_zero]
+    end
+end
```
Good find, will fix.

mmagician commented:
needs a rebase right?

PhilippGackstatter commented on crates/miden-lib/src/account/interface/mod.rs:
```diff
@@ -289,6 +289,27 @@ impl From<&Account> for AccountInterface {
     }
 }
 
+/// Helper function to extract public keys from an account
+pub fn get_public_keys_from_account(account: &Account) -> Vec<Word> {
```
I would not add such free-standing functions to the public `miden-lib` API. I think we should either make it an inherent method on `AccountInterface` or move it behind `#[cfg(test)]` or `testing`.

It seems to be only used for testing purposes in miden-lib and miden-testing, so I would move it to `crates/miden-lib/src/testing/account_interface.rs`.

PhilippGackstatter commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +70,62 @@ proc.assert_new_tx
+        swapw drop push.0
+        # => [[0, 0, 0, i-1], PUB_KEY, i-1]
+
+        push.PUBLIC_KEYS_MAP_SLOT
+        # => [pub_key_slot_idx, [0, 0, 0, i-1], PUB_KEY, i-1]
+
+        exec.account::set_map_item
+        # => [OLD_MAP_ROOT, OLD_MAP_VALUE]
+
+        dropw dropw
+        # => [i-1]
```
```suggestion
        push.PUBLIC_KEYS_MAP_SLOT
        # => [pub_key_slot_idx, [0, 0, 0, i-1], PUB_KEY, i-1]

        exec.account::set_map_item
        # => [OLD_MAP_ROOT, OLD_MAP_VALUE, i-1]

        dropw dropw
        # => [i-1]
```

PhilippGackstatter commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +70,62 @@ proc.assert_new_tx
+    drop drop swap drop
+    # => [num_approvers]
+
+    dup neq.0
+    while.true
+        sub.1
+
```
Nit: Should we add stack comments for this line?

PhilippGackstatter commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +70,62 @@ proc.assert_new_tx
+        dropw dropw
+        # => [i-1]
+
+        dup neq.0
+        # => [is_zero]
```
This stack state also contains the `i` index right?

PhilippGackstatter commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +70,62 @@ proc.assert_new_tx
+    exec.account::set_item
+    # => [OLD_THRESHOLD_CONFIG, NEW_THRESHOLD_CONFIG]
+
+    dropw
+    # => [NEW_THRESHOLD_CONFIG]
+
+    # TODO assert threshold <= num_approvers
+    # TODO only enter loop if pub keys exist in AdviceMap
+
+    drop drop swap drop
+    # => [num_approvers]
```
```suggestion
    dropw
    # => [NEW_THRESHOLD_CONFIG]
    # => [?, ?, num_approvers, ?]

    drop drop swap drop
    # => [num_approvers]
```

Nit: I think this would be more readable if we did something like this, i.e. show the layout of the word on the stack whenever we operate on it directly.

PhilippGackstatter commented on crates/miden-testing/tests/auth/multisig.rs:
```diff
@@ -313,3 +317,153 @@ async fn test_multisig_replay_protection() -> anyhow::Result<()> {
+
+    // for public key in new public keys
+    for (i, public_key) in new_public_keys.iter().enumerate() {
+        let key_word: Word = [Felt::new(i as u64), Felt::new(0), Felt::new(0), Felt::new(1)].into();
```
Can you elaborate why the key conflict is triggered by `EPILOGUE_BEFORE_TX_FEE_REMOVED_FROM_ACCOUNT` (or its event handler?).

I think the larger issue here is that we use a predictable advice map key. Since the advice map is a global namespace for the transaction, using such "well-known" advice map keys seems like a pattern we should not use. I think there is a very good chance another developer would come up with the same idea of using those same keys and if these key-value pairs end up in the same transaction, they would silently overwrite each other which would be pretty annoying to debug and would also make these two account components unusable with each other.

A better solution would probably be for us to expose an event handler that user code can implement for this purpose (e.g. basically this trait https://docs.rs/miden-processor/latest/miden_processor/trait.EventHandler.html). That handler is called by the transaction host when appropriate. We would construct such a handler which contains all the new keys and the multisig MASM code would then instead emit an event when it needs the new keys and the handler would push them onto the advice stack. That would avoid the need to define advice map keys in the first place. I haven't though this through, but something like this seems a bit better than having to choose well-known keys. Of course, in some way, this just shifts the problem to a unique "event ID", but I think the new string-hash approach for events should solve this appropriately.

The alternative is probably to generate advice map keys randomly. Let's say `[1,2,3,4]` was randomly chosen. The multisig component would hardcode that key and then add 1 to a well-specified element to lookup those keys in the map, e.g. `[1,2,3,4]`, `[1,2,3,5]`, `[1,2,3,6]` and so on. That should also work, but feels a bit unclean.

In any case, not something for this PR, but wanted to mention this here.

PhilippGackstatter replied:
I think the best approach for this PR would probably be to:
- Create the vector of public keys of the new signers and hash it. Let's call that hash `SIGNERS_KEY`.
- Insert the vector into the advice map using `SIGNERS_KEY` as the key.
- Pass the `SIGNERS_KEY` to `update_signers_and_threshold` as part of calling it (from the tx script, right?).
- Fetch the vector of new public keys from the advice map using the `SIGNERS_KEY`.

PhilippGackstatter reviewed (commented):
Looks good!

Left a few initial comments.

mmagician commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -123,7 +179,7 @@ export.auth_tx_rpo_falcon512_multisig
     # => [index, TX_SUMMARY_COMMITMENT]
 
-    exec.account::get_item
+    exec.account::get_item_init
```
I think these changes have been merged already?

partylikeits1983 replied:
Not yet :)

#1887 was only to update `rpo_falcon512::verify_signatures`

partylikeits1983 commented on crates/miden-lib/asm/miden/auth/rpo_falcon512.masm:
```diff
@@ -78,6 +78,9 @@ end
 #! from the provided account storage map slot, verifies their signatures against the transaction message,
 #! and returns the number of successfully verified signatures.
 #!
+#! Note: Calls `account::get_map_item_init` to access the transaction’s initial storage state,
+#! enabling validation of updates to the owner public key mapping.
+#!
```
yes, will rebase off of `next` since #1887 got merged

partylikeits1983 commented on crates/miden-testing/tests/auth/multisig.rs:
```diff
@@ -299,17 +302,171 @@ async fn test_multisig_replay_protection() -> anyhow::Result<()> {
+    let sig_1 = authenticators[0].get_signature(public_keys[0].into(), &tx_summary).await?;
+    let sig_2 = authenticators[1].get_signature(public_keys[1].into(), &tx_summary).await?;
+
+    // Execute transaction with signatures - should succeed
```
Great idea, will add this test.

partylikeits1983 replied:
Added the regression test.

partylikeits1983 commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +70,62 @@ proc.assert_new_tx
+    # => [NEW_THRESHOLD_CONFIG]
+
+    # TODO assert threshold <= num_approvers
+    # TODO only enter loop if pub keys exist in AdviceMap
```
Fixed.

partylikeits1983 commented:
> Create the vector of public keys of the new signers and hash it. Let's call that hash SIGNERS_KEY.
Insert the vector into the advice map using SIGNERS_KEY as the key.
Pass the SIGNERS_KEY to update_signers_and_threshold as part of calling it (from the tx script, right?).
Fetch the vector of new public keys from the advice map in update_signers_and_threshold using the SIGNERS_KEY.

@PhilippGackstatter thank you so much for this pointer. Updated this PR to take this exact approach. 

___

@PhilippGackstatter @mmagician Biggest updates since you last reviewed: 

Changed the way how the public keys are passed into the `AdviceMap`. Instead of storing each owner public key at a predictable `KEY` value in the `AdviceMap`, I followed Philipp's advice and created a vector of the new owner public keys and the new multisig config, hashed this vector to get the key, then inserted this vector into the `AdviceMap` under the hash of the vector. 

This is the input to the `update_signers_and_threshold` procedure:

```
Operand stack: [MULTISIG_CONFIG_HASH] - the hash of the threshold and new public key vector
AdviceMap: {
    MULTISIG_CONFIG_HASH => [CONFIG, PUB_KEY_0, PUB_KEY_1, PUB_KEY_N]
}
```
where `CONFIG` is defined as:
```
[threshold, num_approvers, 0, 0]
```

Added two more tests:
1) Test that adds new owner public keys, then validates a new tx using the signatures of the new owners
2) Regression test that Marti suggested, showing that using `account::get_map_item_init` prevents a vulnerability where unauthorized parties could add public keys into the owner pub key mapping, then could authorize the tx themselves since the verify signatures procedure reads from `account:get_map_item`

mmagician commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +72,77 @@ proc.assert_new_tx
+        dup neq.0
+        # => [is_zero, i-1]
+    end
+end
```
> If we have 5 signers initially and then update to 4 signers, the fifth signer will remain in the storage map, right?

That doesn't sound very desirable indeed - is this a limitation that can't be addressed?

partylikeits1983 replied:
It can definitely be addressed. Most likely we need to iterate `old_num_approvers` number of times, and write `EMPTY_WORD` into the storage map to overwrite the old owner. Will also need to add another test to check this works correctly.

PhilippGackstatter replied:
> Will also need to add another test to check this works correctly.

You could also rewrite one of the existing tests to go from 5 to 4 signers and then check that the storage map only has 4 entries. To avoid duplicating a lot of test setup code for this one assertion.

bobbinth commented on crates/miden-lib/src/account/components/mod.rs:
```diff
@@ -82,7 +82,7 @@ pub fn no_auth_library() -> Library {
 }
 
 /// Returns the Multisig Library.
-pub fn multisig_library() -> Library {
+pub fn rpo_falcon_512_multisig_library() -> Library {
```
Let's update the doc comment here as well.

partylikeits1983 replied:
Fixed ! :)

bobbinth commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +72,166 @@ proc.assert_new_tx
+    if.true
+        # => [init_num_of_approvers, new_num_of_approvers]
+
+        dup movup.2 sub dup
+        # => [num_iterations, num_iterations, init_num_of_approvers]
+
+        while.true
```
We could potentially also get rid of the outer `if`-statement. It could look something like this:

```
# => [i = init_num_of_approvers, new_num_of_approvers]

dup.1 dup.1
u32assert2 u32lt
# => [should_loop, i, new_num_of_approvers]

while.true
    # => [i, new_num_of_approvers]

    # remove the approver with index i and then increment i
    # keep looping until i != new_num_of approvers
end
# => [i, new_num_of_approvers]

drop drop
```

partylikeits1983 replied:
Yes, this was an oversight on my part. Using Bobbin's suggestion this is the new flow of the `cleanup_pubkey_mapping` procedure.

bobbinth commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +72,166 @@ proc.assert_new_tx
+#! Update threshold config and add / remove owners
+#!
+#! Inputs:
+#!   Operand stack: [MULTISIG_CONFIG_HASH] - the hash of the threshold and new public key vector
```
I would move the explanation of what `MULTISIG_CONFIG_HASH` into the "where" section (lines 151 - 153 below).

Also, I know we haven't been very consistent with this, but since this procedure is intended to be invoked via the `call` instruction, we should really document padding (both here and in inline comments). So, this would look like:

```
#! Inputs:
#!   Operand stack: [MULTISIG_CONFIG_HASH, pad(12)] 
```

Also also, whey not provide the new `CONFIG` via the stack - i.e., inputscould look like:
```
#! Inputs:
#!   Operand stack: [CONFIG, SIGNERS_HASH, pad(8)]
#!   Advice map: {
#!     SIGNERS_HASH => [PUB_KEY_N, PUB_KEY_N-1, ..., PUB_KEY_0]
#!   }
```

partylikeits1983 replied:
Moved the description of the `MULTISIG_CONFIG_HASH`. Added `pad(12)` to stack comments. 

It would be convenient to pass in the new `CONFIG` via the stack, although this would mean getting the `CONFIG` from the `AdviceStack` in the transaction script which calls the `update_signers_and_threshold` procedure.

bobbinth replied:
Ah yes - good point. I think an alternative could be for the transaction args to be be a key that maps to the combination of config and signers hash - i.e.:
```
TRANSACTION_ARGS |-> [CONFIG, SIGNERS_HASH]
```
And then the advice map for the transaction would look like:
```
Advice map: {
    TRANSACTION_ARGS => [CONFIG, SIGNERS_HASH],
    SIGNERS_HASH => [PUB_KEY_N, PUB_KEY_N-1, ..., PUB_KEY_0],
}
```
But this creates an extra level of indirection, and I'm not sure the benefits are worth it.

partylikeits1983 replied:
> But this creates an extra level of indirection, and I'm not sure the benefits are worth it.

Personally I think this could be a bit confusing 😅

What do you think the main downside of the current layout is?

```
Advice map: {
    MULTISIG_CONFIG_HASH => [CONFIG, PUB_KEY_N, PUB_KEY_N-1, ..., PUB_KEY_0]
}
```

bobbinth reviewed (commented):
Looks good! Thank you. I left a couple of comments inline but agree with @PhilippGackstatter's comments as well.

partylikeits1983 commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +72,166 @@ proc.assert_new_tx
+    dup.1 dup.1
+    # => [init_num_of_approvers, new_num_of_approvers, init_num_of_approvers, new_num_of_approvers]
+
+    lt
```
Fixed this :)

partylikeits1983 commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +72,166 @@ proc.assert_new_tx
+            # => [num_iterations, init_num_of_approvers, num_iterations]
+
+            sub
+            # => [i, num_iterations]
```
Added a test that initialized the multisig with 5 owners, then removes 3 owners.

bobbinth commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +72,166 @@ proc.assert_new_tx
+    drop drop dup movdn.2
+    # => [num_approvers, threshold, num_approvers, pad(12)]
+
+    u32assert2.err=ERR_MALFORMED_MULTISIG_CONFIG u32gt
+    # => [is_invalid_config, pad(12)]
+
+    assertz.err=ERR_MALFORMED_MULTISIG_CONFIG
+    # => [num_approvers, pad(12)]
```
Should the comment here be:
```
# => [is_invalid_config, num_approvers, pad(12)]
```
But also, I maybe just write this as:

```
# make sure that the threshold is smaller than the number of approvers
u32assert2.err=ERR_MALFORMED_MULTISIG_CONFIG
u32gt assertz.err=ERR_MALFORMED_MULTISIG_CONFIG
# => [num_approvers, pad(12)]
```

Also nit: why not check this before we save config to storage? It doesn't make a practical difference, but feels more natural if we first validate the config, and then save it to storage.

----

Separately, it doesn't seem like we ever check if both, the number of approvers and the threshold is greater than 0. I guess this could be a valid use-case where we want to "burn" the account by making sure nobody can use it, but the same could be achieved by having a single approver with a `ZERO` pub key.

So, ensuring that the number of approvers and threshold are always greater than 0 shouldn't reduce functionality while removing a potential footgun.

partylikeits1983 replied:
Makes sense! Refactored to address all these points.

bobbinth commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +72,166 @@ proc.assert_new_tx
+    push.THRESHOLD_CONFIG_SLOT
+    # => [slot, MULTISIG_CONFIG, MULTISIG_CONFIG, pad(12)]
+
+    exec.account::set_item
+    # => [OLD_THRESHOLD_CONFIG, MULTISIG_CONFIG, pad(12)]
```
Since we get the old threshold config here, we could just save the initial number of approvers to memory here (e.g,. using `loc_store.1`), and then we would avoid the need to do `exec.account::get_item_init` on line 209 below.

partylikeits1983 replied:
Ah, that's a really good idea. Refactored to do this!

partylikeits1983 commented on crates/miden-lib/src/account/faucets/network_fungible.rs:
I swear I saw somewhere in the comments of this PR to refactor [crates/miden-lib/src/account/faucets/mod.rs](https://github.com/0xMiden/miden-base/pull/1871/files/d1a8d0fd13fe027a13763452ad3a89103de17a74#diff-d8e7a659e664682ffa37657936c17491c58401d04aaa85d7948d72be6a57f652) into two files, but I seem to have lost this comment in this PR :)

I think I accidentally didn't complete the full refactor of the `mod.rs` file. Can do this in a separate PR since this PR is getting a bit long :)

partylikeits1983 commented on crates/miden-lib/src/testing/account_interface.rs:
```diff
@@ -0,0 +1,28 @@
+    let interface: AccountInterface = account.into();
+
+    for auth in interface.auth() {
+        match auth {
+            AuthScheme::NoAuth => {},
+            AuthScheme::RpoFalcon512 { pub_key } => pub_keys.push(Word::from(*pub_key)),
+            AuthScheme::RpoFalcon512Multisig { pub_keys: multisig_keys, .. } => {
+                for key in multisig_keys {
+                    pub_keys.push(Word::from(*key));
+                }
+            },
+            AuthScheme::Unknown => {},
+        }
```
Created: https://github.com/0xMiden/miden-base/issues/1953

partylikeits1983 commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +72,166 @@ proc.assert_new_tx
+    # => [MULTISIG_CONFIG, pad(12)]
+
+    dropw
+    # => [pad(12)]
```
Refactored to this!

PhilippGackstatter reviewed (approved):
Looks good to me!

Left some minor comments.

partylikeits1983 commented on crates/miden-lib/asm/account_components/multisig_rpo_falcon_512.masm:
```diff
@@ -70,6 +72,155 @@ proc.assert_new_tx
+#! - init_num_of_approvers is the original number of approvers before the update
+#! - new_num_of_approvers is the new number of approvers after the update
+proc.cleanup_pubkey_mapping
+    dup.1 dup.1
+    u32lt
```
Added back the `u32assert`
